crates/wasmparser/src/parser.rs

```diff
@@ -1,1304 +1,1338 @@
+    ///
+    /// This, unlike other sections, is intended to be used for streaming the
+    /// contents of the code section. The code section is not required to be
+    /// fully resident in memory when we parse it. Instead a [`Parser`] is
+    /// capable of parsing piece-by-piece of a code section.
+    CodeSectionStart {
+        /// The number of functions in this section.
         count: u32,
+        /// The range of bytes that represent this section, specified in
+        /// offsets relative to the start of the byte stream.
+        range: Range,
+        /// The size, in bytes, of the remaining contents of this section.
+        ///
+        /// This can be used in combination with [`Parser::skip_section`]
+        /// where the caller will know how many bytes to skip before feeding
+        /// bytes into `Parser` again.
+        size: u32,
     },
-    InstantiateParameter {
-        kind: ExternalKind,
-        index: u32,
-    },
-    EndInstantiate,
-    InlineModule(ModuleCode<'a>),
-}
 
-#[derive(Debug, Copy, Clone)]
-pub enum ElemSectionEntryTable {
-    Passive,
-    Declared,
-    Active(u32),
-}
+    /// An entry of the code section, a function, was parsed.
+    ///
+    /// This entry indicates that a function was successfully received from the
+    /// code section, and the payload here is the window into the original input
+    /// where the function resides. Note that the function itself has not been
+    /// parsed, it's only been outlined. You'll need to process the
+    /// `FunctionBody` provided to test whether it parses and/or is valid.
+    CodeSectionEntry(crate::FunctionBody<'a>),
+
+    /// Indicator of the start of the module code section.
+    ///
+    /// This behaves the same as the `CodeSectionStart` payload being returned.
+    /// You're guaranteed the next `count` items will be of type
+    /// `ModuleCodeSectionEntry`.
+    ModuleCodeSectionStart {
+        /// The number of inline modules in this section.
+        count: u32,
+        /// The range of bytes that represent this section, specified in
+        /// offsets relative to the start of the byte stream.
+        range: Range,
+        /// The size, in bytes, of the remaining contents of this section.
+        size: u32,
+    },
 
-#[derive(Debug, Copy, Clone)]
-pub enum ParserInput {
-    Default,
-    SkipSection,
-    SkipFunctionBody,
-    ReadCustomSection,
-    ReadSectionRawData,
-}
+    /// An entry of the module code section, a module, was parsed.
+    ///
+    /// This variant is special in that it returned a sub-`Parser`. Once a
+    /// `ModuleCodeSectionEntry` it is expected that the returned `Parser` will
+    /// used instead of the parent `Parser` until the parse has finished. You'll
+    /// need to feed data into the `Parser` returned until it returns
+    /// `Payload::End`. After that you'll switch back to the parent parser to
+    /// resume parsing the rest of the module code section.
```

Comment:
```suggestion
    /// This variant is special in that it returns a sub-`Parser`. Upon receiving a
    /// `ModuleCodeSectionEntry` it is expected that the returned `Parser` will be
    /// used instead of the parent `Parser` until the parse has finished. You'll
    /// need to feed data into the `Parser` returned until it returns
    /// `Payload::End`. After that you'll switch back to the parent parser to
    /// resume parsing the rest of the module code section.
```